nifi-nar-bundles/nifi-box-bundle/nifi-box-processors/src/main/java/org/apache/nifi/processors/box/PutBoxFile.java

```diff
@@ -0,0 +1,407 @@
+        return subfolderNames;
+    }
+
+    private BoxFolder.Info getOrCreateSubfolders(Queue<String> subFolderNames, BoxFolder parentFolder, boolean createFolder) {
+        final BoxFolder.Info folderInfo = getOrCreateFolder(subFolderNames.poll(), parentFolder, createFolder);
+
+        if (!subFolderNames.isEmpty()) {
+           final BoxFolder newParentFolder = getFolder(folderInfo.getID());
+           return getOrCreateSubfolders(subFolderNames, newParentFolder, createFolder);
+        } else {
+            return folderInfo;
+        }
+    }
+
+    private BoxFolder.Info getOrCreateFolder(String folderName, BoxFolder parentFolder, boolean createFolder) {
+        final Optional<BoxFolder.Info> existingFolder = getFolderByName(folderName, parentFolder);
+
+        if (existingFolder.isPresent()) {
+            return existingFolder.get();
+        }
+
+        if (!createFolder) {
+           throw new ProcessException(format("The specified subfolder [%s] does not exist and [%s] is false.",
+                   folderName, CREATE_SUBFOLDER.getDisplayName()));
+        }
+
+        return createFolder(folderName, parentFolder);
+    }
+
+    private Info createFolder(final String folderName, final BoxFolder parentFolder) {
+        getLogger().info("Creating Folder [{}], Parent [{}]", folderName, parentFolder.getInfo().getID());
+
+        try {
+           return parentFolder.createFolder(folderName);
+        } catch (BoxAPIResponseException e) {
+            getLogger().info("Folder [{}], Parent [{}] already existed", folderName, parentFolder.getInfo().getID());
+
+            if (e.getResponseCode() != CONFLICT_RESPONSE_CODE) {
+                throw e;
+            } else {
+                return getFolderByName(folderName, parentFolder).orElseThrow(() ->
+                        new ProcessException(format("Created subfolder [%s] can not be found under [%s]", folderName, parentFolder.getInfo().getID()), e));
+            }
+        }
+    }
+
+    private BoxFolder getFolderById(final String folderId) {
+        final BoxFolder folder = getFolder(folderId);
+        try {
+            //Error is returned for nonexistent folder only when a method is called on BoxFolder.
+            folder.getInfo();
+        } catch (BoxAPIResponseException e) {
+            if (e.getResponseCode() == NOT_FOUND_RESPONSE_CODE) {
+                throw new ProcessException(format("The Folder [%s] specified by [%s] does not exist", folderId, FOLDER_ID.getDisplayName()));
+            }
+        }
+        return folder;
+    }
+
+    private Optional<BoxFolder.Info> getFolderByName(final String folderName, final BoxFolder parentFolder) {
+        return getItemByName(folderName, parentFolder, BoxFolder.Info.class);
+    }
+
+    private Optional<BoxFile.Info> getFileByName(final String filename, final BoxFolder parentFolder) {
+        return getItemByName(filename, parentFolder, BoxFile.Info.class);
+    }
```

Comment:
Converting `BoxFolder.Info` to `BoxFolder` and then back appear in the code in several places.
I might consider converting the Info objects to BoxFolder/BoxFile right here, and use those objects in the rest of the code.
```
    private Optional<BoxFile> getFileByName(final String filename, final BoxFolder parentFolder) {
        return getItemByName(filename, parentFolder, BoxFile.Info.class).map(BoxFile.Info::getResource);
    }
```